pypeit/pypeit.py

```diff
@@ -718,7 +718,8 @@ def save_exposure(self, frame, all_spec2d, all_specobjs, basename):
                                        slitspatnum=self.par['rdx']['slitspatnum'])
             # Info
             outfiletxt = os.path.join(self.science_path, 'spec1d_{:s}.txt'.format(basename))
-            all_specobjs.write_info(outfiletxt, self.spectrograph.pypeline)
+            sobjs = specobjs.SpecObjs.from_fitsfile(outfile1d, chk_version=False)
```

Comment:
This does not seem like the correct approach to me. I see the problem that you are trying to solve but the solution you propose seems highly sub-optimal. Only a small fraction of reductions will reduce the detectors separately. Nevertheless, you are constantly re-reading the sobjs from disk here and passing to the routine that writes the spec1d.txt file. I think the better approach would be simply to append the writes to the spec1d.txt file, rather than creating a new file each time. In other words, if that file already exists, you simply append to it, if that file does not exist yet you create it. This would obviate the need to continually re-read the spec1d outputs from the disk, and rewrite the spec1d file.

Reply:
I think I prefer this new way.

But as this was related to one of @marlageha 's requests, 
I defer to her preference.

Reply:
The reason I used this approach is to deal with the following scenario: you run the reduction for all the detectors, but then find out that the reduction for one detector was not good and you want to re-run it only for that detector.

If we append to the existing spec1d.txt file, it will lists both the old extractions (that one that are not good) and the new ones, creating some confusion. With this new approach, the spec1d.txt will always lists only the latest extraction information for each reduced detector.

Reply:
I  don't like re-reading the entire spec1d file from disk every time just to modify a few lines in a text file for a corner usage case. 

Also,  this syntax is opaque and in my opinion silly. Just append sobjs to all_specobjs. Why should this write method also pass a new sobjs as an argument. 

all_specobjs.write_info(outfiletxt, self.spectrograph.pypeline, sobjs=sobjs)

Maybe it is time to consider generating separate spec1d files for each detector. 

Otherwise, my suggestion is that we simply at the every end of the entire reduction or re-reduction regenerate the spec*.txt file by reading in the spec1d object and running this method once. 

Joe

Reply:
> I think I prefer this new way.
> 
> But as this was related to one of @marlageha 's requests,
> I defer to her preference.

I am no longer using the spec1d*txt files, once it was pointed out that the spec1d*fits headers have all of the *txt information.   That said, they are useful for quick checks.   I agree with Joe's last comment above about regenerating spec1d*txt at the end of each redux/re-redux.

Reply:
@jhennawi I want to make absolute sure that this is also clear to you: this method is not run for every detector, i.e., the spec1d datamodel is not read from disk for every detector or multiple times during the reduction. This is alway done once at the end of the reduction (or re-reduction). Even the spec1d datamodel itself is written to disk at the very end of the reduction process.  

So, if you run the full reduction for all 8 DEIMOS detectors, at the very end PypeIt writes the `spec1d.fits` file that includes all the detectors. Then, if you decide to re-run the reduction only for detector 1 (for example), at the very end of the reduction the `spec1d.fits` file is updated with the new extractions for the detector 1, while the other detectors remain untouched.  This is why it made sense to me to create the `spec1.txt` file by reading the `spec1d.fits` at the end of the reduction.

About the syntax, I agree with you. It is not great. I was trying make sure I was not affecting other pieces of code. But I can definitely append  `sobjs` to `all_specobjs` and remove `sobjs` as an argument.